libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp

```diff
@@ -257,6 +272,28 @@ void AP_InertialSensor_Backend::apply_gyro_filters(const uint8_t instance, const
     } else {
         _imu._gyro_filtered[instance] = gyro_filtered;
     }
+
+#if AP_INERTIALSENSOR_RATE_LOOP_WINDOW_ENABLED
+    if (_imu.push_rate_loop_gyro(instance)) {
+        /*
+          tell the rate thread we have a new sample
+        */
+        if (++_imu.rate_decimation_count >= _imu.rate_decimation) {
+            _imu._cmutex->lock_and_signal();
```

Comment:
why are you signalling before you push?? I think this may be why you are getting the odd binary sem behaviour. You have to signal after you have made the sample available or the rate thread will wake up pointlessly and you will always be one loop behind

Reply:
This makes no difference - the waiter will not be woken up until it gets the mutex which won't happen until the mutex is unlocked by the signaler.

If the waiter is in chCondWait the first thing it needs to do is re-aquire the mutex which it cannot do until the signaler has released it.

https://chibios.org/dokuwiki/lib/exe/fetch.php?media=chibios:documentation:books:rt:kernel:monitor_states.png

As the docs state this is very similar to java synchronized which I am very familiar with:

https://chibios.org/dokuwiki/doku.php?id=chibios:documentation:books:rt:kernel_mutexes